[fix] [broker] Fix broker deadlock in metadata-callback thread due to single thread processing

rdhabalia opened the pull request:
Fixes https://github.com/apache/pulsar/issues/22840

<!-- or this PR is one task of an issue -->

Main Issue: #22840

### Motivation

This PR addresses the first issue discussed in #22840
Having a single metadata store callback can easily cause deadlock in any regression. Therefore, we must have multiple or the same number of callback threads as number of IO threads so that, callback thread doesn't face bottleneck and doesn't cause any deadlock in the system.

```
"metadata-store-10-1" #25 prio=5 os_prio=0 cpu=7574.38ms elapsed=703.54s tid=0x00007fafdd884ed0 nid=0x37eb waiting for monitor entry  [0x00007fae8f9d3000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(java.base@11.0.19/ConcurrentHashMap.java:1723)
        - waiting to lock <0x00001000f4a02bc0> (a java.util.concurrent.ConcurrentHashMap$ReservationNode)
        at org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl.asyncOpen(ManagedLedgerFactoryImpl.java:365)
        at org.apache.pulsar.broker.service.BrokerService.lambda$createPersistentTopic$66(BrokerService.java:1533)
        at org.apache.pulsar.broker.service.BrokerService$$Lambda$644/0x00007facb91c8040.accept(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniAcceptNow(java.base@11.0.19/CompletableFuture.java:753)
        at java.util.concurrent.CompletableFuture.uniAcceptStage(java.base@11.0.19/CompletableFuture.java:731)
        at java.util.concurrent.CompletableFuture.thenAccept(java.base@11.0.19/CompletableFuture.java:2108)
        at org.apache.pulsar.broker.service.BrokerService.createPersistentTopic(BrokerService.java:1514)
        at org.apache.pulsar.broker.service.BrokerService.lambda$checkOwnershipAndCreatePersistentTopic$60(BrokerService.java:1472)
        at org.apache.pulsar.broker.service.BrokerService$$Lambda$634/0x00007facb91c6968.accept(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniAcceptNow(java.base@11.0.19/CompletableFuture.java:753)
        at java.util.concurrent.CompletableFuture.uniAcceptStage(java.base@11.0.19/CompletableFuture.java:731)
        at java.util.concurrent.CompletableFuture.thenAccept(java.base@11.0.19/CompletableFuture.java:2108)
        at org.apache.pulsar.broker.service.BrokerService.checkOwnershipAndCreatePersistentTopic(BrokerService.java:1470)
        at org.apache.pulsar.broker.service.BrokerService.lambda$loadOrCreatePersistentTopic$57(BrokerService.java:1437)
        at org.apache.pulsar.broker.service.BrokerService$$Lambda$492/0x00007facc317a960.run(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(java.base@11.0.19/CompletableFuture.java:783)
        at java.util.concurrent.CompletableFuture.postComplete(java.base@11.0.19/CompletableFuture.java:506)
        at java.util.concurrent.CompletableFuture.complete(java.base@11.0.19/CompletableFuture.java:2073)
        at org.apache.pulsar.metadata.coordination.impl.LockManagerImpl.lambda$acquireLock$1(LockManagerImpl.java:105)
        at org.apache.pulsar.metadata.coordination.impl.LockManagerImpl$$Lambda$613/0x00007facbf1d9d08.run(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(java.base@11.0.19/CompletableFuture.java:783)
        at java.util.concurrent.CompletableFuture.postComplete(java.base@11.0.19/CompletableFuture.java:506)
        at java.util.concurrent.CompletableFuture.complete(java.base@11.0.19/CompletableFuture.java:2073)
        at org.apache.pulsar.metadata.coordination.impl.ResourceLockImpl.lambda$acquire$2(ResourceLockImpl.java:128)
        at org.apache.pulsar.metadata.coordination.impl.ResourceLockImpl$$Lambda$611/0x00007facbdc40040.run(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(java.base@11.0.19/CompletableFuture.java:783)
        at java.util.concurrent.CompletableFuture.postComplete(java.base@11.0.19/CompletableFuture.java:506)
        at java.util.concurrent.CompletableFuture.complete(java.base@11.0.19/CompletableFuture.java:2073)
        at org.apache.pulsar.metadata.coordination.impl.ResourceLockImpl.lambda$acquireWithNoRevalidation$6(ResourceLockImpl.java:167)
        at org.apache.pulsar.metadata.coordination.impl.ResourceLockImpl$$Lambda$609/0x00007facbdc42908.accept(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniAccept.tryFire(java.base@11.0.19/CompletableFuture.java:714)
        at java.util.concurrent.CompletableFuture.postComplete(java.base@11.0.19/CompletableFuture.java:506)
        at java.util.concurrent.CompletableFuture.complete(java.base@11.0.19/CompletableFuture.java:2073)
        at org.apache.pulsar.metadata.impl.ZKMetadataStore.handlePutResult(ZKMetadataStore.java:225)
        at org.apache.pulsar.metadata.impl.ZKMetadataStore.lambda$batchOperation$7(ZKMetadataStore.java:182)
        at org.apache.pulsar.metadata.impl.ZKMetadataStore$$Lambda$160/0x00007fae8db950b0.run(Unknown Source)
        at java.util.concurrent.Executors$RunnableAdapter.call(java.base@11.0.19/Executors.java:515)
        at java.util.concurrent.FutureTask.run(java.base@11.0.19/FutureTask.java:264)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(java.base@11.0.19/ScheduledThreadPoolExecutor.java:304)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.19/ThreadPoolExecutor.java:1128)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.19/ThreadPoolExecutor.java:628)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(java.base@11.0.19/Thread.java:829)




"pulsar-io-4-32" #166 prio=5 os_prio=0 cpu=429.42ms elapsed=701.77s tid=0x00007facc80314b0 nid=0x3881 waiting on condition  [0x00007facc73f4000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@11.0.19/Native Method)
        - parking to wait for  <0x00001000f838ac60> (a java.util.concurrent.CompletableFuture$Signaller)
        at java.util.concurrent.locks.LockSupport.park(java.base@11.0.19/LockSupport.java:194)
        at java.util.concurrent.CompletableFuture$Signaller.block(java.base@11.0.19/CompletableFuture.java:1796)
        at java.util.concurrent.ForkJoinPool.managedBlock(java.base@11.0.19/ForkJoinPool.java:3128)
        at java.util.concurrent.CompletableFuture.waitingGet(java.base@11.0.19/CompletableFuture.java:1823)
        at java.util.concurrent.CompletableFuture.get(java.base@11.0.19/CompletableFuture.java:1998)
        at org.apache.pulsar.bookie.rackawareness.BookieRackAffinityMapping.setConf(BookieRackAffinityMapping.java:121)
        - locked <0x00001000f8389698> (a org.apache.pulsar.bookie.rackawareness.BookieRackAffinityMapping)
        at org.apache.bookkeeper.client.RackawareEnsemblePlacementPolicyImpl.initialize(RackawareEnsemblePlacementPolicyImpl.java:265)
        at org.apache.pulsar.bookie.rackawareness.IsolatedBookieEnsemblePlacementPolicy.initialize(IsolatedBookieEnsemblePlacementPolicy.java:105)
        at org.apache.pulsar.bookie.rackawareness.IsolatedBookieEnsemblePlacementPolicy.initialize(IsolatedBookieEnsemblePlacementPolicy.java:51)
        at org.apache.bookkeeper.client.BookKeeper.initializeEnsemblePlacementPolicy(BookKeeper.java:581)
        at org.apache.bookkeeper.client.BookKeeper.<init>(BookKeeper.java:505)
        at org.apache.bookkeeper.client.BookKeeper$Builder.build(BookKeeper.java:306)
        at org.apache.pulsar.broker.BookKeeperClientFactoryImpl.create(BookKeeperClientFactoryImpl.java:87)
        at org.apache.pulsar.broker.ManagedLedgerClientFactory.lambda$initialize$0(ManagedLedgerClientFactory.java:95)
        at org.apache.pulsar.broker.ManagedLedgerClientFactory$$Lambda$1006/0x00007facb41b1440.apply(Unknown Source)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(java.base@11.0.19/ConcurrentHashMap.java:1705)
        - locked <0x00001000f3c03a28> (a java.util.concurrent.ConcurrentHashMap$ReservationNode)
        at org.apache.pulsar.broker.ManagedLedgerClientFactory.lambda$initialize$1(ManagedLedgerClientFactory.java:93)
        at org.apache.pulsar.broker.ManagedLedgerClientFactory$$Lambda$202/0x00007fae1e73e9b0.get(Unknown Source)
        at org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl.lambda$asyncOpen$6(ManagedLedgerFactoryImpl.java:369)
        at org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl$$Lambda$660/0x00007facb91ecd60.apply(Unknown Source)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(java.base@11.0.19/ConcurrentHashMap.java:1705)
        - locked <0x00001000f4a02bc0> (a java.util.concurrent.ConcurrentHashMap$ReservationNode)
        at org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl.asyncOpen(ManagedLedgerFactoryImpl.java:365)
        at org.apache.pulsar.broker.service.BrokerService.lambda$createPersistentTopic$66(BrokerService.java:1533)
        at org.apache.pulsar.broker.service.BrokerService$$Lambda$644/0x00007facb91c8040.accept(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniAcceptNow(java.base@11.0.19/CompletableFuture.java:753)
        at java.util.concurrent.CompletableFuture.uniAcceptStage(java.base@11.0.19/CompletableFuture.java:731)
        at java.util.concurrent.CompletableFuture.thenAccept(java.base@11.0.19/CompletableFuture.java:2108)
        at org.apache.pulsar.broker.service.BrokerService.createPersistentTopic(BrokerService.java:1514)
        at org.apache.pulsar.broker.service.BrokerService.lambda$checkOwnershipAndCreatePersistentTopic$60(BrokerService.java:1472)
        at org.apache.pulsar.broker.service.BrokerService$$Lambda$634/0x00007facb91c6968.accept(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniAcceptNow(java.base@11.0.19/CompletableFuture.java:753)
        at java.util.concurrent.CompletableFuture.uniAcceptStage(java.base@11.0.19/CompletableFuture.java:731)
        at java.util.concurrent.CompletableFuture.thenAccept(java.base@11.0.19/CompletableFuture.java:2108)
        at org.apache.pulsar.broker.service.BrokerService.checkOwnershipAndCreatePersistentTopic(BrokerService.java:1470)
        at org.apache.pulsar.broker.service.BrokerService.lambda$loadOrCreatePersistentTopic$57(BrokerService.java:1437)
        at org.apache.pulsar.broker.service.BrokerService$$Lambda$492/0x00007facc317a960.run(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniRunNow(java.base@11.0.19/CompletableFuture.java:815)
        at java.util.concurrent.CompletableFuture.uniRunStage(java.base@11.0.19/CompletableFuture.java:799)
        at java.util.concurrent.CompletableFuture.thenRun(java.base@11.0.19/CompletableFuture.java:2121)
        at org.apache.pulsar.broker.service.BrokerService.loadOrCreatePersistentTopic(BrokerService.java:1433)
        at org.apache.pulsar.broker.service.BrokerService.lambda$getTopic$31(BrokerService.java:1014)
        at org.apache.pulsar.broker.service.BrokerService$$Lambda$476/0x00007facc3057108.apply(Unknown Source)
        at org.apache.pulsar.common.util.collections.ConcurrentOpenHashMap$Section.put(ConcurrentOpenHashMap.java:404)
        at org.apache.pulsar.common.util.collections.ConcurrentOpenHashMap.computeIfAbsent(ConcurrentOpenHashMap.java:238)
        at org.apache.pulsar.broker.service.BrokerService.getTopic(BrokerService.java:1013)
        at org.apache.pulsar.broker.service.BrokerService.getTopic(BrokerService.java:978)
        at org.apache.pulsar.broker.service.BrokerService.lambda$getOrCreateTopic$29(BrokerService.java:973)
```

This PR has unit-test which fails with single thread and it can be fixed by providing multiple threads into callback processing.


### Modifications

Allow metadata-store to use multiple callback threads to avoid deadlock situation in broker.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

PR in forked repository: <!-- ENTER URL HERE -->

<!--
After opening this PR, the build in apache/pulsar will fail and instructions will
be provided for opening a PR in the PR author's forked repository.

apache/pulsar pull requests should be first tested in your own fork since the 
apache/pulsar CI based on GitHub Actions has constrained resources and quota.
GitHub Actions provides separate quota for pull requests that are executed in 
a forked repository.

The tests will be run in the forked repository until all PR review comments have
been handled, the tests pass and the PR is approved by a reviewer.
-->

lhotari commented:
My comment on a similar previous PR: https://github.com/apache/pulsar/pull/22542#issuecomment-2067142518

merlimat commented:
I don't think this is caused by having a single-thread processing the metadata-store callbacks (which should be typically solved by avoiding mixing sync/async calls). 

In this case it looks just a deadlock in the `BookieRackAffinityMapping` code. The blocking of the metadata-callback thread is purely a side effect of the `ManagedLedgerFactoryImpl` map lock being held by the other thread. 

It would be useful to see which thread is holding that lock: 

```
        - parking to wait for  <0x00001000f838ac60> (a java.util.concurrent.CompletableFuture$Signaller)
```

Can you search for it in the full stack trace?

merlimat commented:
@rdhabalia One more question, which code version are these stack traces based on?

rdhabalia commented:
Hello @merlimat 
> I don't think this is caused by having a single-thread processing the metadata-store callbacks (which should be typically solved by avoiding mixing sync/async calls).

That's correct but it's difficult to avoid such mixing if it comes as a part of any regression or new development. Currently it's happening due to sync call at `BookieRackAffinityMapping` which blocking-call I tried to fix in #22842 and #22843 . but this issue can be also mitigated if we would have multiple metadata-callback threads and I tried to simulate it in unit-case as well where unit-test case goes in deadlock and test fails with single metadata thread callback.

> It would be useful to see which thread is holding that lock:
> - parking to wait for  <0x00001000f838ac60> (a java.util.concurrent.CompletableFuture$Signaller)


no one is holding `0x00001000f838ac60` because that's where [BookieRackAffinityMapping](https://github.com/apache/pulsar/blob/branch-3.3/pulsar-broker-common/src/main/java/org/apache/pulsar/bookie/rackawareness/BookieRackAffinityMapping.java#L122) waiting for the `CompletableFuture` to be completed. Please find [full jstack for more info](https://gist.github.com/rdhabalia/f50c7b7b86a78427194603072d3b7666)
```
 at jdk.internal.misc.Unsafe.park(java.base@11.0.19/Native Method)
        - parking to wait for  <0x00001000f838ac60> (a java.util.concurrent.CompletableFuture$Signaller)
        at java.util.concurrent.locks.LockSupport.park(java.base@11.0.19/LockSupport.java:194)
        at java.util.concurrent.CompletableFuture$Signaller.block(java.base@11.0.19/CompletableFuture.java:1796)
        at java.util.concurrent.ForkJoinPool.managedBlock(java.base@11.0.19/ForkJoinPool.java:3128)
        at java.util.concurrent.CompletableFuture.waitingGet(java.base@11.0.19/CompletableFuture.java:1823)
        at java.util.concurrent.CompletableFuture.get(java.base@11.0.19/CompletableFuture.java:1998)
        at org.apache.pulsar.bookie.rackawareness.BookieRackAffinityMapping.setConf(BookieRackAffinityMapping.java:121)
```


> In this case it looks just a deadlock in the BookieRackAffinityMapping code. The blocking of the metadata-callback thread is purely a side effect of the ManagedLedgerFactoryImpl map lock being held by the other thread.

well, overall issue was:: metadata-store callback thread is blocked on topic creation, and other thread is blocked on different topic creation which is waiting on metadata-result's completableFuture and that will be only completed if metadata-store callback gets free. and that creates a deadlock.

Please find below stacktrace for better understanding:

1. `metadata-store` thread was trying to create ledger and held a lock into `ManagedLedgerFactoryImpl` 
```
"metadata-store-10-1" #25 prio=5 os_prio=0 cpu=7574.38ms elapsed=703.54s tid=0x00007fafdd884ed0 nid=0x37eb waiting for monitor entry  [0x00007fae8f9d3000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(java.base@11.0.19/ConcurrentHashMap.java:1723)
	- waiting to lock <0x00001000f4a02bc0> (a java.util.concurrent.ConcurrentHashMap$ReservationNode)
	at org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl.asyncOpen(ManagedLedgerFactoryImpl.java:365)
```

2. At the same time, other thread `pulsar-io` was also trying to create a new topic which required `ManagedLedgerClientFactory` to create a new BK-Client and bk-client has a blocking call on metadatastore to get rack-info and waiting on CompletableFuture of metadatastore which is not being completed as metadatastore has only 1 thread and its waiting on step-1.
```
"pulsar-io-4-32" #166 prio=5 os_prio=0 cpu=429.42ms elapsed=701.77s tid=0x00007facc80314b0 nid=0x3881 waiting on condition  [0x00007facc73f4000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@11.0.19/Native Method)
        - parking to wait for  <0x00001000f838ac60> (a java.util.concurrent.CompletableFuture$Signaller)
        at java.util.concurrent.locks.LockSupport.park(java.base@11.0.19/LockSupport.java:194)
        at java.util.concurrent.CompletableFuture$Signaller.block(java.base@11.0.19/CompletableFuture.java:1796)
        at java.util.concurrent.ForkJoinPool.managedBlock(java.base@11.0.19/ForkJoinPool.java:3128)
        at java.util.concurrent.CompletableFuture.waitingGet(java.base@11.0.19/CompletableFuture.java:1823)
        at java.util.concurrent.CompletableFuture.get(java.base@11.0.19/CompletableFuture.java:1998)
        at org.apache.pulsar.bookie.rackawareness.BookieRackAffinityMapping.setConf(BookieRackAffinityMapping.java:121)
        - locked <0x00001000f8389698> (a org.apache.pulsar.bookie.rackawareness.BookieRackAffinityMapping)
        at org.apache.bookkeeper.client.RackawareEnsemblePlacementPolicyImpl.initialize(RackawareEnsemblePlacementPolicyImpl.java:265)
        at org.apache.pulsar.bookie.rackawareness.IsolatedBookieEnsemblePlacementPolicy.initialize(IsolatedBookieEnsemblePlacementPolicy.java:105)
        at org.apache.pulsar.bookie.rackawareness.IsolatedBookieEnsemblePlacementPolicy.initialize(IsolatedBookieEnsemblePlacementPolicy.java:51)
        at org.apache.bookkeeper.client.BookKeeper.initializeEnsemblePlacementPolicy(BookKeeper.java:581)
        at org.apache.bookkeeper.client.BookKeeper.<init>(BookKeeper.java:505)
        at org.apache.bookkeeper.client.BookKeeper$Builder.build(BookKeeper.java:306)
        at org.apache.pulsar.broker.BookKeeperClientFactoryImpl.create(BookKeeperClientFactoryImpl.java:87)
        at org.apache.pulsar.broker.ManagedLedgerClientFactory.lambda$initialize$0(ManagedLedgerClientFactory.java:95)
        at org.apache.pulsar.broker.ManagedLedgerClientFactory$$Lambda$1006/0x00007facb41b1440.apply(Unknown Source)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(java.base@11.0.19/ConcurrentHashMap.java:1705)
        - locked <0x00001000f3c03a28> (a java.util.concurrent.ConcurrentHashMap$ReservationNode)
        at org.apache.pulsar.broker.ManagedLedgerClientFactory.lambda$initialize$1(ManagedLedgerClientFactory.java:93)
        at org.apache.pulsar.broker.ManagedLedgerClientFactory$$Lambda$202/0x00007fae1e73e9b0.get(Unknown Source)
        at org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl.lambda$asyncOpen$6(ManagedLedgerFactoryImpl.java:369)
        at org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl$$Lambda$660/0x00007facb91ecd60.apply(Unknown Source)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(java.base@11.0.19/ConcurrentHashMap.java:1705)
        - locked <0x00001000f4a02bc0> (a java.util.concurrent.ConcurrentHashMap$ReservationNode)
        at org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl.asyncOpen(ManagedLedgerFactoryImpl.java:365)
```


I have added unit-test which simulate the similar behavior. Our brokers were keep going in this deadlock situation and becoming irresponsive. Therefore, we deployed this quick fix and this issue didn't happen again. Therefore, I have created this fix to mitigate any future regression issue and I have also fixed mixing of sync/async call in other PRs.

Attached stacktrace is from broker-2.10 but it happens with broker-3.3 as well.

merlimat commented:
> but this issue can be also mitigated if we would have multiple metadata-callback threads and I tried to simulate it in unit-case as well where unit-test case goes in deadlock and test fails with single metadata thread callback.

I don't think that it would be a mitigation of the issue. If you have N threads you can still have the same issue depending on which threads the requests are assigned. 

The only solution is to not do blocking operation where these are not expected, unfortunately.

rdhabalia commented:
> If you have N threads you can still have the same issue depending on which threads the requests are assigned. 

It might not give a guarantee for all scenarios but it helps in most of the scenarios to prevent the deadlock. Before Metadaastore implementation, we had multiple zk callback threads to avoid this issue which was working in most of the cases.
we are wasting lot of time to keep our system stable with such issues. and this fix helps us to keep it stable in such regressions where extra thread helps to prevent race-condition and because of that avoid the deadlock.

I don't see any harm to keep it multi threads for metadata-store and if you don't want to keep in all usecases then I can make it configurable from service configuration.

merlimat commented:
> It might not give a guarantee for all scenarios 

That's equal to no guarantees

> I don't see any harm to keep it multi threads for metadata-store and if you don't want to keep in all usecases then I can make it configurable from service configuration.

It does bring multiple problems: 
 1. Masks problems that must be eradicated at the source. The problem is still there, and it will still be happening, just it becomes harder to track.
 2. Multiple threads will bring in ordering issue across callbacks
 3. Potentially more mutex contention

> we are wasting lot of time to keep our system stable with such issues. and this fix helps us to keep it stable in such regressions where extra thread helps to prevent race-condition and because of that avoid the deadlock.

I understand, though the problem was introduced by masking the blocking call in the interface you introduced a while back.

rdhabalia commented:
> I understand, though the problem was introduced by masking the blocking call in the interface you introduced a while back.

umm.. which interface? The issue is the interface in BookKeeper client which doesn't support async call and I don't recall any such interface in BookKeeper introduced by me.

also, if we use ordered executor then it will not cause ordering. and anyway metadata calling thread and metadata callback threads will be always different so, I don't think it makes much difference for context switching.

I agree we should improve mixing sync/async code path but there is also not any harm in supporting multiple callback threads model as it's not anything new which we are proposing.

merlimat commented:
>umm.. which interface? The issue is the interface in BookKeeper client which doesn't support async call and I don't recall any such interface in BookKeeper introduced by me.

`BookkeeperFactoryForCustomEnsemblePlacementPolicy`

Sure, the main issue is in BK client, though this issue is only present when using different BK clients based on namespace policy. 

This is an example of problematic code, try/catch masking a blocking operation: 
https://github.com/apache/pulsar/blob/342d88dd193bb85c0af91c5193b1422808a9c821/pulsar-broker/src/main/java/org/apache/pulsar/broker/ManagedLedgerClientFactory.java#L100-L109

That also breaks the strict bookie assignment requirement that you were mentioning, because it will fallback to default BK client if there is any failure to read the bookie mapping from metadata. What's worse, it will not self-correct, and it will continue to use the wrong BK client indefinitely.

```java
// find or create bk-client in cache for a specific ensemblePlacementPolicy
            if (ensemblePlacementPolicyConfig != null && ensemblePlacementPolicyConfig.getPolicyClass() != null) {
                bkClient = bkEnsemblePolicyToBkClientMap.computeIfAbsent(ensemblePlacementPolicyConfig, (key) -> {
                    try {
                        return bookkeeperProvider.create(conf, metadataStore, eventLoopGroup,
                                Optional.ofNullable(ensemblePlacementPolicyConfig.getPolicyClass()),
                                ensemblePlacementPolicyConfig.getProperties(), statsLogger);
                    } catch (Exception e) {
                        log.error("Failed to initialize bk-client for policy {}, properties {}",
                                ensemblePlacementPolicyConfig.getPolicyClass(),
                                ensemblePlacementPolicyConfig.getProperties(), e);
                    }
                    return this.defaultBkClient;
                });
            }
```

rdhabalia commented:
I don't see any issue with this PR but I am not going to waste my time as other PRs are opened and I don't think one wants to review it. so, please feel free to close.
